Relabel log levels, higher level is more important

pgdr opened the pull request:
**Task**
Fix #211

Update log levels according to more conventional design where a _higher number_ means _more important_ message.

Specifically, we set

| Level   | Value | Use  |
|---      |---    |---   |
| DEBUG   | 10    | Detailed information, typically of interest only when diagnosing problems.     |
| INFO    | 20    | Confirmation that things are working as expected.     |
| WARNING | 30    | An indication that something unexpected happened, or indicative of some problem in the near future (e.g. ‘disk space low’). The software is still working as expected.     |
| ERROR   | 40    |   Due to a more serious problem, the software has not been able to perform some function.   |
| CRITICAL| 50    |   A serious error, indicating that the program itself may be unable to continue running.   |



**Approach**
In the enum, `LOG_CRITICAL=0` and I suspect that somewhere in libres there is a
```c
if (log_level)
  log_open(logh);
```

Commit:
Refactor the log levels

Relabel log levels, higher level is more important
Pass message_level_enum into Python
Remove function from API and remove arg from func
Remove argument free_message from (res_)log_add_message_str, never free
Remove function log_add_message_py, use plain log_add_message_str
Add translate log level to test
Add res_log_info, debug, etc, make msg const
Remove unused symbols is_open and get_logh
Add shortcut functions for fmt print log messages

joakim-hove commented on libres_util/include/ert/res_util/log.h:
```diff
@@ -29,31 +29,47 @@ extern "C" {
-  int          log_get_level( const log_type * logh);
-  void         log_set_level( log_type * logh , int new_level);
+  void         log_add_fmt_message(log_type * logh , message_level_type message_level , FILE * dup_stream , const char * fmt , ...);
+  message_level_type          log_get_level( const log_type * logh);
```
Please maintain the indentation which is currently in the code.

pgdr replied:
I can reindent the rest of the code if that is what we want.  Aligned code is a double-edged sword.

markusdregi replied:
I'm in favor of not reindenting, for the exact reason displayed here. Header files are too crucial to program flow that it is worth the cost of reindenting vs. the loss of git history in my opinion.

If you end up not reindenting, could you at least make the ~10 white spaces into a single one?

joakim-hove commented on libres_util/src/res_log.c:
```diff
@@ -112,24 +110,31 @@ void res_log_add_message(message_level_type message_level , FILE * dup_stream ,
-      log_add_message( logh , false , NULL , "Exiting ert application normally - all is fine(?)" , false);
+void res_log_close() {
+    if (log_is_open(logh))
+      log_add_message(logh,
```
I agree that many of the lines in the code are too long, and appreciate the splitting; in particular I was not aware of the possibility to split string literals like `"part1" "part2"` - that was very good to learn. However splitting the current `log_add_message( )` call to cover four vertical lines is a bit excessive in my opinion.

pgdr replied:
> However splitting [...]  four vertical lines is a bit excessive in my opinion.

*If the arguments fit, I will acquit.*

However, I like to _either_ keep all arguments on one line, _or_ split all lines.  I am open for suggestions.

joakim-hove replied:
Well - this is not important; but I would prefer the current function call to be one *one single line*

> I like to either keep all arguments on one line, or split all lines.

My take on this is to try to group related arguments  - when that makes sense - on the same line. 

Anyway - no big deal either way.

pgdr replied:
fixed

joakim-hove commented on python/python/res/util/enums/message_level_enum.py:
```diff
@@ -9,10 +9,39 @@ class MessageLevelEnum(BaseCEnum):
+    @staticmethod
+    def to_enum(val):
+        if val is None:
+            return MessageLevelEnum.LOG_WARNING
```
Do we envisage actually using a numerical log level of e.g. 33?

pgdr replied:
Well, the classical loggers like log4j, log4c and Python's logging all accept integer levels for logging.  I don't know if we want that though.

joakim-hove commented on python/python/res/util/res_log.py:
```diff
@@ -1,19 +1,26 @@
+    @staticmethod
+    def __enum(log_level):
+        if isinstance(log_level, int):
+            log_level = MessageLevelEnum.to_enum(log_level)
```
What to do if the `isinstance(log_level, int)` test fails - exception?

pgdr replied:
This was actually just under the assumption that the type is either int or a MessageLevelEnum.

joakim-hove replied:
I guess that is why it is slightly dangerous to start with `isinstance( )` checks in Python, it kind of breaks the duck-typing contract. Maybe the `MessageLevelEnum.to_enum( )` could be modified to accept a enum instance as input argument, then that method can be called unconditionally?

joakim-hove commented on python/python/res/util/res_log.py:
```diff
@@ -1,19 +1,26 @@
     def log(cls, log_level, message):
         """ Higher log_level indicates higher importance"""
-        cls._write_log(log_level, message)
+        cls._write_log(cls.__enum(log_level), message)
```
Wasn't it natural that the `log( )` method now expected an enum argument?

pgdr replied:
Yes, but now it supports both int and MessageLevelEnum.  I'm open for suggestions here!

joakim-hove reviewed (commented):
It seems the Python api is now set up to use a mix of the new enum and integer log levels - or do I misunderstand? Apart from that this looks good.

Commit:
Remove log_reopen from API, move log file in snake

Commit:
Rename default log file to log.txt

Commit:
Use res_log_xxx

Commit:
Remove several methods from res_log API

Update tests after new default log level
Remove __enum and raise if not MessageLevel

markusdregi reviewed (approved):
Reviewed in private communication. Merge when green.

Commit:
Revert lines broken at col 80
